### PR TITLE
BUIP055 partial implementation

### PR DESCRIPTION
**Pull request by gandrewstone**
Implements:
fork trigger
opt-in replay protection for 1MB chain txs

### CODE REVIEW

**Comment by ptschip** [comment]
needs a rebase

File: doc/buip-hf.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+BUIP-HF Development Notes
```
Comment (ftrader) [inline doc/buip-hf.md:1]: Most of this document was applicable to working on my personal repo.

I think this document does not need to remain, or can be trimmed down do bare essentials as the work continues on BU's `dev` branch.

File: qa/rpc-tests/test_framework/util.py [hunk 2]
```diff
@@ -150,28 +151,34 @@ def hex_str_to_bytes(hex_str):
...
-def sync_mempools(rpc_connections, wait=1):
+def sync_mempools(rpc_connections, wait=1,verbose=1):
     """
     Wait until everybody has the same transactions in their memory
     pools
     """
     while True:
         pool = set(rpc_connections[0].getrawmempool())
         num_match = 1
+        poolLen = [len(pool)]
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/util.py:174]: nit pick: why using camelcase here?

File: src/main.h [hunk 3]
```diff
@@ -113,7 +113,7 @@ static const int64_t BLOCK_DOWNLOAD_TIMEOUT_BASE = 1000000;
 /** Additional block download timeout per parallel downloading peer (i.e. 5 min) */
 static const int64_t BLOCK_DOWNLOAD_TIMEOUT_PER_PEER = 500000;
 /** Timeout in secs for the initial sync. If we don't receive the first batch of headers */
-static const uint32_t INITIAL_HEADERS_TIMEOUT = 30;
+static const uint32_t INITIAL_HEADERS_TIMEOUT = 120;
```
Comment (sickpig) [inline src/main.h:116]: why is this change needed?

File: src/buip055fork.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        if (!ValidateBUIP055Tx(tx))
+            return state.DoS(100,
+                             error("transaction is invalid on BUIP055 chain"), REJECT_INVALID, "bad-txns-wrong-fork");
+    }
+    return true;
+}
+
+bool ValidateBUIP055Tx(const CTransaction& tx)
+{
```
Comment (sickpig) [inline src/buip055fork.cpp:83]: I cant's see anywhere the code that implement the sunset rule that consider this kind of txn invalid only if block height <= 530000, see REQ-6.1.

File: src/buip055fork.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    return false;
+}
+
+bool ValidateBUIP055Block(const CBlock &block, CValidationState &state)
+{
+    // Validate transactions are HF compatible
+    for (const CTransaction &tx : block.vtx)
+    {
+        if (!ValidateBUIP055Tx(tx))
+            return state.DoS(100,
```
Comment (sickpig) [inline src/buip055fork.cpp]: Why using 100, that means that the peer will be dos banned as soon as we see a txn that activate the replay protection, is this really what we want?

File: src/buip055fork.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+#include <boost/foreach.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/thread.hpp>
+#include <inttypes.h>
+#include <iomanip>
+#include <limits>
+#include <queue>
+
+std::vector<unsigned char> invalidOpReturn = {112, 101, 101, 114, 50, 112, 101, 101, 114, 32, 99, 97, 115, 104, 32, 114, 101, 113, 117, 105, 114, 101, 115, 32, 108, 97, 114, 103, 101, 114, 32, 98, 108, 111, 99, 107, 115};
+
```
Comment (sickpig) [inline src/buip055fork.cpp]: why not just using an `static const std::string` ("peer2peer cash requires larger blocks") rather than a vector of unsigned char and convert the it to a vector just before the comparison in `IsTxOpReturnInvalid`? 
IMHO having `invalidOpReturn` stated as a string would make the code clearer.

File: src/main.h (hunk 3 above)
Comment (gandrewstone) [reply src/main.h:116]: buip055 test was failing because of timeout on my test setup, when I used a rotating disk.  But working fine using a ramdisk.  I talked to @ptschip and we agreed that its ok to increase this value.

File: src/buip055fork.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/buip055fork.cpp]: in this case, a peer has given us a block with a bad tx in it.  The tx has replay protection, but it should not.  So it has mis-coded the BUIP055 rules.  I suppose we could not ban it, just ignore the block...

File: src/main.h (hunk 3 above)
Comment (sickpig) [reply src/main.h:116]: mmm ok, was just worried by the fact that the change does not seem related to the PR

File: src/buip055fork.cpp (hunk 5 above)
Comment (sickpig) [reply src/buip055fork.cpp]: no strong opinion, I wonder what other think? paging @ftrader @ptschip @deadalnix @dagurval @dgenr8

File: src/main.cpp [hunk 7]
```diff
@@ -1154,6 +1155,16 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
     if (tx.IsCoinBase())
         return state.DoS(100, false, REJECT_INVALID, "coinbase");
 
+    // BUIP055: reject transactions that won't work on the fork.
+    // This code uses the system time to determine when to start rejecting which is inaccurate relative to the
+    // actual activation time (defined by times in the blocks).
```
Comment (dgenr8) [inline src/main.cpp:1162]: Could chainActive.Tip()->GetMedianTimePast() be used?

File: src/main.cpp [hunk 8]
```diff
@@ -4042,6 +4054,26 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
...
+            return state.DoS(100,
+                error("%s: BUIP055 fork block (%s, height %d) must exceed %d, but this block is %d bytes", __func__,
+                                 hash.ToString(), nHeight, BLOCKSTREAM_CORE_MAX_BLOCK_SIZE, block.nBlockSize),
+                REJECT_INVALID, "bad-blk-too-small");
+        }
+    }
+    // BUIP055 check soft-fork items, such as tx targeted to the 1MB chain
+    if (pindexPrev && pindexPrev->IsforkActiveOnNextBlock(miningForkTime.value))
+    {
+        return ValidateBUIP055Block(block, state);
```
Comment (dgenr8) [inline src/main.cpp]: Could this be done inside the existing tx loop in ContextualCheckBlock, rather than a new tx loop?

**Comment by sickpig** [comment]
@gandrewstone this PR need a rebase

File: src/buip055fork.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017 The Bitcoin Unlimited Developers
```
Comment (sickpig) [inline src/buip055fork.cpp:1]: wonder if we should call it `uahffork.cpp` or `bitcashfork.cpp`

File: src/buip055fork.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2017 The Bitcoin Unlimited Developers
```
Comment (sickpig) [inline src/buip055fork.h:1]: same here we should consider to rename the file
